XP boosters

drac-dot-gay opened the pull request:
this pr includes XP boosters and related logic

Commit:
updated/refactored user booster entity

Commit:
updated/refactored queued booster entity

Commit:
updated/refactored available slot entity

Commit:
updated/refactored active booster entity

Commit:
added boosterservice

Commit:
converted boosterservice into commands

Commit:
added updateboostercyclecommand

Commit:
fixed xp unit tests

Commit:
fixed some activateboostercommand exceptions

Commit:
fixed addboostercommand exceptions

Commit:
made getboostermultipliercommand synchronous

Commit:
fixed updateboostercyclecommand exceptions

Commit:
fixed populateboosterservicecommand exceptions

Commit:
added guild channel rewriting

Commit:
add userbooster commands and queries

Commit:
wired up events and background jobs

Commit:
added adduserboostercommand

Commit:
added GetActiveBoostersQuery

Commit:
added GetQueuedBoostersQuery

Commit:
added GetAvailableSlotsQuery

Commit:
moved slot commands (organization yay)

Commit:
QueuedBooster bugfix

(WHY CANT IT HAVE A PRIMARY KEY BRUH)

Commit:
commandmediator modification

Commit:
added boostercommandmodule

Commit:
housekeeping!

Commit:
added activateboostercommandtests

Commit:
added PopulateBoosterServiceCommandTests

Commit:
added GetBoosterMultiplierCommandTests

Commit:
fixed a bug where GetBoosterMultiplierCommand would throw an error if you tried to access a guild that didn't exist

thanks, unit tests!

Commit:
added SaveQueueCommandTests

TGrannen committed:
Fix float calculation issue with boosters

TGrannen committed:
Add User level up event handler to create user boosters

TGrannen committed:
Handle ghost level ups when applying level roles to user

TGrannen committed:
Add evil twin message sent in channel. Also refactor some XP and Level events to pass new model around instead of properties on each event

TGrannen committed:
Fix multiple user boosters if user levels up many levels

TGrannen committed:
Remove unneeded creation of XP user record until user gains xp

Commit:
updated nunit test adapter

Commit:
added UpdateBoosterCycleCommand tests

Commit:
updated get booster multiplier logic

boosters are no longer calculated as 1 + sum(boosters), instead as either 1 or the sum of any existing boosters.

Commit:
brought down the number of tests used for getmultiplier

69 is still a funny number though

Commit:
removed GetUserBoosterQuery

Commit:
fixed cron expression for booster cycle updates

Commit:
weeee local functions

TGrannen commented on GarbageCan.Application/Boosters/Commands/UpdateBoosterCycleCommand.cs:
```diff
@@ -0,0 +1,87 @@
+
+            while (_boosterService.QueuedBoosters[request.GuildId].Count > 0 &&
+                   _boosterService.ActiveBoosters[request.GuildId].Count <
+                   _boosterService.AvailableSlots[request.GuildId].Count)
```
I feel like there's a better solution to this. Like a query that just gets the Boosters and Slots for a single guild so you don't have to do this all over the place? Might not be worth it at this point if it's functional. Just feel like it's kinda convoluted

drac-dot-gay replied:
i pushed my idea for a fix in commit de59ea5d930893fd44a725e6562146ba46eca978

TGrannen commented on GarbageCan.Web/Commands/Boosters/BoosterCommandModule.cs:
```diff
@@ -0,0 +1,289 @@
+        [Command("use")]
+        public async Task UseBooster(CommandContext ctx, int id)
+        {
+            try
```
This whole method should probably be a whole command in itself like `UseBoosterCommand`

TGrannen commented on GarbageCan.Application/XP/Services/XpCalculatorService.cs:
```diff
@@ -1,24 +1,37 @@
+            float GetMultiplier() =>
+                _mediator.Send(new GetBoosterMultiplierCommand { GuildId = guildId })
+                    .GetAwaiter()
+                    .GetResult();
```
If this is going to do an async operation, it'd be better to expose this as an async method than try to do this synchronously

drac-dot-gay replied:
i figured this was probably best or at least not a huge deal because if you look at the command being called its synchronous

TGrannen commented on GarbageCan.Application/XP/Services/XpCalculatorService.cs:
```diff
@@ -1,24 +1,37 @@
+
+        public double XpEarned(string message, ulong guildId)
+        {
+            float GetMultiplier() =>
```
Doesn't look like there's really a need for a local method here. Just more code to read. The mediator call can be just used down on line 29 instead.

Commit:
tweaked boosterresult handling slightly

this counteracts a potential bug where if something goes wrong and the command somehow returns something its not supposed to, it doesn't just ignore it and remove the specified booster

Commit:
added suggested solution to the review comment in PR #15

TGrannen committed:
Add example of restructure command response sending

TGrannen committed:
Removing models for responding and just using parameters instead

Commit:
made xp calculator async

Commit:
fixed a entityframework core extensions related bug

Commit:
updated populateboosterservicecommand to fix an issue of constant exceptions if a server has no defined slots

Commit:
test hotfix so these dont immediately fail

TGrannen committed:
Restructuring all the responses to be within the Command Modules instead of the handlers

TGrannen committed:
Merge remote-tracking branch 'origin/feature/boosters' into feature/CommandResponseRestructure

# Conflicts:
#	GarbageCan.Application/Common/Interfaces/IDiscordGuildService.cs
#	GarbageCan.Application/XP/Queries/GetUserLevelImage/GetUserLevelImageQuery.cs
#	GarbageCan.Infrastructure/Discord/DiscordGuildService.cs

Commit:
Merge pull request #16 from murl-digital/feature/CommandResponseRestructure

Command restructuring

Commit:
behold; my janky test solution

TGrannen committed:
Simplify populate booster service tests
